validator/lib.go

```diff
@@ -1,10 +1,559 @@
+// Validator must provide a `Verifier` corresponding to local authority.
+//
+// A capability provider service will use one corresponding to own DID or it's
+// supervisor's DID if it acts under it's authority.
+//
+// This allows a service identified by non did:key e.g. did:web or did:dns to
+// pass resolved key so it does not need to be resolved at runtime.
+type Validator interface {
+	Authority() principal.Verifier
+}
+
+type ClaimContext interface {
+	Validator
+	RevocationChecker[any]
+	CanIssuer[any]
+	ProofResolver
+	PrincipalParser
+	PrincipalResolver
+}
+
+type ValidationContext[Caveats any] interface {
+	ClaimContext
+	Capability() CapabilityParser[Caveats]
+}
+
+type validationContext[Caveats any] struct {
+	authority             principal.Verifier
+	capability            CapabilityParser[Caveats]
+	canIssue              CanIssueFunc[any]
+	validateAuthorization RevocationCheckerFunc[any]
+	resolveProof          ProofResolverFunc
+	parsePrincipal        PrincipalParserFunc
+	resolveDIDKey         PrincipalResolverFunc
+}
+
+func (vc validationContext[Caveats]) Authority() principal.Verifier {
+	return vc.authority
+}
+
+func (vc validationContext[Caveats]) Capability() CapabilityParser[Caveats] {
+	return vc.capability
+}
+
+func (vc validationContext[Caveats]) CanIssue(capability ucan.Capability[any], issuer did.DID) bool {
+	return vc.canIssue(capability, issuer)
+}
+
+func (vc validationContext[Caveats]) ValidateAuthorization(auth Authorization[any]) result.Result[result.Unit, Revoked] {
+	return vc.validateAuthorization(auth)
+}
+
+func (vc validationContext[Caveats]) ResolveProof(proof ucan.Link) result.Result[delegation.Delegation, UnavailableProof] {
+	return vc.resolveProof(proof)
+}
+
+func (vc validationContext[Caveats]) ParsePrincipal(str string) (principal.Verifier, error) {
+	return vc.parsePrincipal(str)
+}
+
+func (vc validationContext[Caveats]) ResolveDIDKey(did did.DID) result.Result[did.DID, UnresolvedDID] {
+	return vc.resolveDIDKey(did)
+}
+
+func NewValidationContext[Caveats any](
+	authority principal.Verifier,
+	capability CapabilityParser[Caveats],
+	canIssue CanIssueFunc[any],
+	validateAuthorization RevocationCheckerFunc[any],
+	resolveProof ProofResolverFunc,
+	parsePrincipal PrincipalParserFunc,
+	resolveDIDKey PrincipalResolverFunc,
+) ValidationContext[Caveats] {
+	return validationContext[Caveats]{
+		authority,
+		capability,
+		canIssue,
+		validateAuthorization,
+		resolveProof,
+		parsePrincipal,
+		resolveDIDKey,
+	}
+}
+
+// Access finds a valid path in a proof chain of the given `invocation` by
+// exploring every possible option. On success an `Authorization` object is
+// returned that illustrates the valid path. If no valid path is found
+// `Unauthorized` error is returned detailing all explored paths and where they
+// proved to fail.
+func Access[Caveats any](invocation invocation.Invocation, context ValidationContext[Caveats]) (result.Result[Authorization[Caveats], Unauthorized], error) {
```

Comment:
Tracing these functions all the way back to the service, we we understand well the difference between returning a result.Error and a golang error? I see there is a HandlerExecutionError but I also find this whole distinction a bit cryptic.

I just want to make sure this makes sense. By default I would expect function returning a result type not to return an error. I'm also ok though if it makes sense.

Reply:
Ahh, yes at this point I'm not sure we're getting loads of benefit from returning a result type - it was useful when first porting the code but considering Go has typed errors in return values I think I'd be open to just using that as it's more ideomatic, and then wrap up a "Result" for the wire and use them in the receipts.

Reply:
honestly this would be my suggestion